Address external debug issue #11 (#20)
riscv/riscv-smmtt: chapter8.adoc

[[chapter8]]
[[Smsdedbg]]
== `Smsdedbg`: Supervisor Domain External Debug

[caption="Figure {counter:image}: ", reftext="Figure {image}"]
[title= "External Debug for Supervisor Domain", id=Smsdedbg_img]
image::Smsdedbg.png[]

The <<MSDCFG>> CSR contains the `sdedbgen` bit that controls whether the
current scheduled SD is external-debug enabled. This bit is context switched

> (along with rest of the `msdcfg`) per SD. The `medbgen` is an enable control

Reviewer: The "enabled" terminology could be avoided as the DSM cannot determine whether external debug is enabled. It may be better to use "whether external debug is allowed for the currently scheduled SD".

Author: updated

for external debug for the M-mode driven by the debug module and is expected
to be established by the RoT (following RISC-V Security Model recommendation

> SR_GEN_007 and SR_GEN_012). When privilege is `M`, the `medbgen` gates the

Reviewer 2: We need to separate the concepts between Smmtt Spec and External debug security Spec. My suggestion would be to move the content after this line to External debug security Spec. The separation would be following: 
- Smmtt Spec: Definition of medbgen, sdedbgen, MSDCFG, Smsdedbg ISA Extension
- External debug security Spec: The rules about how external debug and trace behaves based on medbgen, sdedbgen.
I am open to your suggestion for the separation. My goal is to have single source of truth.

Author: agree with your proposal @gokhankaplayan - i have separated the paragraph that line onwards and marked as as Informative for this spec (to be moved into the external debug security spec.

`haltreq_i` from the debug module and if is 0 prevents the hart from entering
external debug mode. When privilege is less than `M`, the OR of the 
`MSDCFG.sdedbgen` and `medbgen` gates the `haltreq_i` from the debug module
and the hart will enter external debug mode if either is 1.

The configuration for `MSDCFG.sdedbgen` is obtained from the manifest of the
SD and is managed by the M-mode root security manager using secure memory.

When `medbgen` is 1, there are no restrictions. When `medbgen` is 0 but
`MSDCFG.sdedbgen` is 1, then the external debug mode may be entered but has
to have restrictions such as `M-mode`-privileged CSR accesses are prohibited
and Memory accesses by instructions executed in external debug mode cannot
use `M` privilege. When `medbgen` is 0 and `MSDCFG.sdedbgen` is also 0, then

> the rules around triggers configured with action = 1 that may fire - such as

Reviewer 2: In external debug security draft spec, the approach is for Triggers with action = 1. It is not allowed to configure them with action = 1 depending on tdata1.{m,s,u} and debug enables. If they fire as breakpoint exception, it might be a potential security leak (denial of service, privilege escalation attack?).

Reviewer: Triggers can be controlled by the RDSM. They will be context switched so this control is only needed for halting external debug.

always cause breakpoint exception. The external debug security extension TG

> will need to define the rules around what is allowed in external debug mode.

Reviewer: The ISA specification may not want to get so specific about the non-ISA aspect. This text should get separated out into a corresponding non-ISA specification. That would avoid a tight coupling between what the ISA specification says - "External debug is allowed" and what the non-ISA specification says "How".

Reviewer: The triggers dont need any special consideration. They are context switched by the RDSM.

[[chapter8]]
[[Smsdedbg]]
== `Smsdedbg`: Supervisor Domain External Debug

[caption="Figure {counter:image}: ", reftext="Figure {image}"]
[title= "External Debug for Supervisor Domain", id=Smsdedbg_img]
image::Smsdedbg.png[]

> 

Reviewer: The picture can move into the section to be moved to external debug security.

> [caption="Figure {counter:image}: ", reftext="Figure {image}"]
> [title= "External Debug for Supervisor Domain", id=Smsdedbg_img]
> image::Smsdedbg.png[]
> 

Author: ```suggestion
```

The <<MSDCFG>> CSR contains the `sdedbgalw` bit that controls whether the
current scheduled SD is allowed to be external-debugged. This bit is context
switched (along with rest of the `msdcfg`) per SD.

> 

Reviewer: The last sentence may be somewhat redundant.

=== `Smsdedbg` interaction with external debug security controls (Informative)

> 

Author: ```suggestion

[caption="Figure {counter:image}: ", reftext="Figure {image}"]
[title= "External Debug for Supervisor Domain", id=Smsdedbg_img]
image::Smsdedbg.png[]

```

This section will be moved into the non-ISA specification for external debug
security. It is described here as informational.

The `medbgen` is an enable control for external debug for the M-mode driven by
the debug module and is expected to be established by the RoT (following RISC-V
Security Model recommendation SR_GEN_007 and SR_GEN_012). When privilege is `M`,

> the `medbgen` gates the `haltreq_i` from the debug module and if is 0 prevents

Author: ```suggestion
the `medbgen` gates the `haltreq` from the debug module and if is 0 prevents
```

> the hart from entering external debug mode. When privilege is less than `M`, the

Reviewer: haltreq_i -> haltreq

> OR of the `MSDCFG.sdedbgalw` and `medbgen` gates the `haltreq_i` from the debug

Author: ```suggestion
OR of the `MSDCFG.sdedbgalw` and `medbgen` gates the `haltreq` from the debug
```

module and the hart will enter external debug mode if either is 1.

> The configuration for `MSDCFG.sdedbgalw` may be obtained from the manifest/
> configuration of the supervisoer domain and should be managed by the M-mode root
> security manager using secure memory.

Author: ```suggestion
The configuration for `MSDCFG.sdedbgalw` may be obtained from the manifest/
configuration of the supervisor domain and should be managed by the M-mode root
security manager using memory accessible only to the M-mode root security manager.
```

When `medbgen` is 1, there are no restrictions. When `medbgen` is 0 but
`MSDCFG.sdedbgalw` is 1, then the external debug mode may be entered but has
to have restrictions such as `M-mode` -privileged CSR accesses are prohibited
and Memory accesses by instructions executed in external debug mode cannot
use `M` privilege. When `medbgen` is 0 and `MSDCFG.sdedbgalw` is also 0, then

> the rules around triggers configured with action = 1 that may fire - such as

Reviewer: I dont think this is necessary. when `sdebgalw` is 0, the RDSM should just not configure triggers with `action=1`. I dont think we need to this logic to hardware.

> When `medbgen` is 1, there are no restrictions. When `medbgen` is 0 but
> `MSDCFG.sdedbgalw` is 1, then the external debug mode may be entered but has
> to have restrictions such as `M-mode` -privileged CSR accesses are prohibited
> and Memory accesses by instructions executed in external debug mode cannot
> use `M` privilege. When `medbgen` is 0 and `MSDCFG.sdedbgalw` is also 0, then
> the rules around triggers configured with action = 1 that may fire - such as
> always cause breakpoint exception. The external debug security extension TG
> will need to define the rules around what is allowed in external debug mode.

Author: ```suggestion
When `medbgen` is 1, there are no restrictions. When `medbgen` is 0 but
`MSDCFG.sdedbgalw` is 1, then the external debug mode may be entered but is
restricted by `M-mode` software to prevent privileged CSR accesses and memory
accesses by instructions executed in external debug mode cannot use `M` privilege.
When `medbgen` is 0 and `MSDCFG.sdedbgalw` is also 0, when sdebgalw is 0, the
M-mode root domain security manager must not configure triggers with `action=1`.
```
